pennylane_lightning/core/src/simulators/lightning_kokkos/measurements/MeasurementsKokkos.hpp

```diff
@@ -649,13 +650,16 @@ class Measurements final
      * Reference https://en.wikipedia.org/wiki/Inverse_transform_sampling
      *
      * @param num_samples Number of Samples
+     * @param seed Seed to generate the samples from
      *
      * @return std::vector<std::size_t> to the samples.
      * Each sample has a length equal to the number of qubits. Each sample can
      * be accessed using the stride sample_id*num_qubits, where sample_id is a
      * number between 0 and num_samples-1.
      */
-    auto generate_samples(std::size_t num_samples) -> std::vector<std::size_t> {
+    auto generate_samples(std::size_t num_samples,
+                          const std::optional<std::size_t> &seed = std::nullopt)
```

Comment:
Since the contained type is a POD, do we need the `optional` to be a constant reference, or will a value be sufficient?

Reply:
The issue with a default argument value is, what would the value be? It should not be zero since zero is a valid seed value (seed is unsigned 32bit int), so the default argument value would be -1. This makes the argument to be a type "bigger" than `uint32_t`, and the argument type would mismatch its purpose a bit.

Reply:
> This makes the argument to be a type "bigger" than `uint32_t`

Let's call this "bigger type" `int`, for illustration's sake.

This leads to the following problem: the seed argument would need to have type `int` just for the default value of -1. Some other piece of code interfacing with this function (say, e.g., some other rng is generating the seed to pass in here) wouldn't know that the actual underlying expected type is `size_t`, and could potentially just pass in -1 (or -2, ...). This is bad because:
- a user can pass in -1, expecting a seeded run, and getting an unseeded run (since that was the default value we would check against)
- [std::mt19937 has undefined behavior if it's not initialized with unsigned](https://en.cppreference.com/w/cpp/numeric/random/mersenne_twister_engine#:~:text=The%20effect%20is%20undefined%20if%20this%20is%20not%20one%20of%20unsigned%20short%2C%20unsigned%20int%2C%20unsigned%20long%2C%20or%20unsigned%20long%20long.)

These problems would not be present if we explicitly specify it as `std::optional<size_t>`, avoiding the default value / signed-ness problem altogether.

(Also, having this argument as `size_t` will make sure the underlying rng is initialized with an unsigned value.)

Reply:
@mlxd I suddenly realized you were talking about `const std::optional<size_t> &seed` vs `std::optional<size_t> seed, not vs `int seed = -1` :sweat_smile: 

I don't think there's that much of a difference whether it's a const reference or not. I would suggest keep it as is just so we don't have to wait for CI again : ))))

Reply:
Sure, fine by me ;)